File: graphlib-2/ts/src/graph.ts
// Graph class and basic operations

import {
  Graph,
  GraphOptions,
  CreateGraphResult,
  AddNodeResult,
  AddEdgeResult,
  RemoveNodeResult,
  RemoveEdgeResult,
  GetNodesResult,
  GetEdgesResult,
  GetNeighborsResult,
  HasNodeResult,
  HasEdgeResult,
  GetDegreeResult,
  GetGraphInfoResult,
  ClearGraphResult,
  CloneGraphResult,
  SubgraphResult,
  Edge,
} from './types';
import { getNextId, storeGraph, loadGraph } from './storage';

export function getGraph(graphId: string): Graph | undefined {
  return loadGraph(graphId);
}

export function createGraph(options: GraphOptions = {}): CreateGraphResult {
  const id = getNextId();
  const graph: Graph = {
    id,
    directed: options.directed ?? false,
    weighted: options.weighted ?? false,
    nodes: new Set(),
    adjacency: new Map(),
    inEdges: new Map(),
  };
  storeGraph(graph);
  return {
    id,
    directed: graph.directed,
    weighted: graph.weighted,
    node_count: 0,
    edge_count: 0,
  };
}

export function addNode(graphId: string, nodeId: string): AddNodeResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { success: false, error: 'graph_not_found' };
  }
  if (graph.nodes.has(nodeId)) {
    return { success: false, error: 'node_already_exists' };
  }
  graph.nodes.add(nodeId);
  graph.adjacency.set(nodeId, new Map());
  graph.inEdges.set(nodeId, new Map());
  storeGraph(graph);
  return { success: true, node_id: nodeId };
}

export function addEdge(
  graphId: string,
  fromNode: string,
  toNode: string,
  weight?: number
): AddEdgeResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { success: false, error: 'graph_not_found' };
  }

  // Auto-create nodes if they don't exist
  if (!graph.nodes.has(fromNode)) {
    graph.nodes.add(fromNode);
    graph.adjacency.set(fromNode, new Map());
    graph.inEdges.set(fromNode, new Map());
  }
  if (!graph.nodes.has(toNode)) {
    graph.nodes.add(toNode);
    graph.adjacency.set(toNode, new Map());
    graph.inEdges.set(toNode, new Map());
  }

  const edgeWeight = weight ?? 1.0;

  // Check if edge already exists
  if (graph.adjacency.get(fromNode)!.has(toNode)) {
    return { success: false, error: 'edge_already_exists' };
  }

  // Add edge
  graph.adjacency.get(fromNode)!.set(toNode, edgeWeight);
  graph.inEdges.get(toNode)!.set(fromNode, edgeWeight);

  // For undirected graphs, add reverse edge
  if (!graph.directed) {
    graph.adjacency.get(toNode)!.set(fromNode, edgeWeight);
    graph.inEdges.get(fromNode)!.set(toNode, edgeWeight);
  }

  storeGraph(graph);
  return { success: true, from: fromNode, to: toNode, weight: edgeWeight };
}

export function removeNode(graphId: string, nodeId: string): RemoveNodeResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { success: false, error: 'graph_not_found' };
  }
  if (!graph.nodes.has(nodeId)) {
    return { success: false, error: 'node_not_found' };
  }

  let removedEdges = 0;

  // Count outgoing edges from this node
  const outgoing = graph.adjacency.get(nodeId);
  if (outgoing) {
    removedEdges += outgoing.size;
    // Remove from inEdges of target nodes
    for (const target of outgoing.keys()) {
      graph.inEdges.get(target)?.delete(nodeId);
    }
  }

  // Count incoming edges to this node
  const incoming = graph.inEdges.get(nodeId);
  if (incoming) {
    // For directed graphs, count separately; for undirected, already counted
    if (graph.directed) {
      removedEdges += incoming.size;
    }
    // Remove from adjacency of source nodes
    for (const source of incoming.keys()) {
      graph.adjacency.get(source)?.delete(nodeId);
    }
  }

  // Remove the node itself
  graph.nodes.delete(nodeId);
  graph.adjacency.delete(nodeId);
  graph.inEdges.delete(nodeId);

  storeGraph(graph);
  return { success: true, removed_edges: removedEdges };
}

export function removeEdge(
  graphId: string,
  fromNode: string,
  toNode: string
): RemoveEdgeResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { success: false, error: 'graph_not_found' };
  }

  const fromAdj = graph.adjacency.get(fromNode);
  if (!fromAdj || !fromAdj.has(toNode)) {
    return { success: false, error: 'edge_not_found' };
  }

  // Remove edge
  fromAdj.delete(toNode);
  graph.inEdges.get(toNode)?.delete(fromNode);

  // For undirected, remove reverse edge
  if (!graph.directed) {
    graph.adjacency.get(toNode)?.delete(fromNode);
    graph.inEdges.get(fromNode)?.delete(toNode);
  }

  storeGraph(graph);
  return { success: true };
}

export function getNodes(graphId: string): GetNodesResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { nodes: [], count: 0 };
  }
  const nodes = Array.from(graph.nodes).sort();
  return { nodes, count: nodes.length };
}

export function getEdges(graphId: string): GetEdgesResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { edges: [], count: 0 };
  }

  const edges: Edge[] = [];
  const seen = new Set<string>();

  for (const [from, neighbors] of graph.adjacency) {
    for (const [to, weight] of neighbors) {
      const key = graph.directed ? `${from}->${to}` : [from, to].sort().join('-');
      if (!seen.has(key)) {
        seen.add(key);
        edges.push({ from, to, weight });
      }
    }
  }

  return { edges, count: edges.length };
}

export function getNeighbors(graphId: string, nodeId: string): GetNeighborsResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { neighbors: [], count: 0 };
  }

  const adj = graph.adjacency.get(nodeId);
  if (!adj) {
    return { neighbors: [], count: 0 };
  }

  const neighbors = Array.from(adj.keys()).sort();
  return { neighbors, count: neighbors.length };
}

export function hasNode(graphId: string, nodeId: string): HasNodeResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { exists: false };
  }
  return { exists: graph.nodes.has(nodeId) };
}

export function hasEdge(
  graphId: string,
  fromNode: string,
  toNode: string
): HasEdgeResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { exists: false };
  }

  const adj = graph.adjacency.get(fromNode);
  if (!adj || !adj.has(toNode)) {
    return { exists: false };
  }

  return { exists: true, weight: adj.get(toNode) };
}

export function getDegree(graphId: string, nodeId: string): GetDegreeResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { degree: 0, in_degree: 0, out_degree: 0 };
  }

  const outDegree = graph.adjacency.get(nodeId)?.size ?? 0;
  const inDegree = graph.inEdges.get(nodeId)?.size ?? 0;

  if (graph.directed) {
    return {
      degree: outDegree + inDegree,
      in_degree: inDegree,
      out_degree: outDegree,
    };
  } else {
    // For undirected, in_degree = out_degree = degree
    return {
      degree: outDegree,
      in_degree: outDegree,
      out_degree: outDegree,
    };
  }
}

export function getGraphInfo(graphId: string): GetGraphInfoResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return {
      id: graphId,
      directed: false,
      weighted: false,
      node_count: 0,
      edge_count: 0,
      is_connected: true,
      has_cycle: false,
    };
  }

  const edgesResult = getEdges(graphId);

  // Inline connectivity check to avoid circular dependency
  let isConn = true;
  if (graph.nodes.size > 1) {
    const visited = new Set<string>();
    const startNode = graph.nodes.values().next().value as string;
    const queue: string[] = [startNode];
    visited.add(startNode);
    while (queue.length > 0) {
      const current = queue.shift()!;
      const outNeighbors = graph.adjacency.get(current) ?? new Map();
      const inNeighbors = graph.inEdges.get(current) ?? new Map();
      const allNeighbors = new Set([...outNeighbors.keys(), ...inNeighbors.keys()]);
      for (const neighbor of allNeighbors) {
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          queue.push(neighbor);
        }
      }
    }
    isConn = visited.size === graph.nodes.size;
  }

  // Inline cycle check
  let hasCycleResult = false;
  if (graph.nodes.size > 0) {
    if (graph.directed) {
      const WHITE = 0, GRAY = 1, BLACK = 2;
      const color = new Map<string, number>();
      for (const node of graph.nodes) color.set(node, WHITE);
      const dfs = (node: string): boolean => {
        color.set(node, GRAY);
        for (const neighbor of graph.adjacency.get(node)?.keys() ?? []) {
          if (color.get(neighbor) === GRAY) return true;
          if (color.get(neighbor) === WHITE && dfs(neighbor)) return true;
        }
        color.set(node, BLACK);
        return false;
      };
      for (const node of graph.nodes) {
        if (color.get(node) === WHITE && dfs(node)) {
          hasCycleResult = true;
          break;
        }
      }
    } else {
      const visited = new Set<string>();
      const dfs = (node: string, parent: string | null): boolean => {
        visited.add(node);
        for (const neighbor of graph.adjacency.get(node)?.keys() ?? []) {
          if (!visited.has(neighbor)) {
            if (dfs(neighbor, node)) return true;
          } else if (neighbor !== parent) {
            return true;
          }
        }
        return false;
      };
      for (const node of graph.nodes) {
        if (!visited.has(node) && dfs(node, null)) {
          hasCycleResult = true;
          break;
        }
      }
    }
  }

  return {
    id: graph.id,
    directed: graph.directed,
    weighted: graph.weighted,
    node_count: graph.nodes.size,
    edge_count: edgesResult.count,
    is_connected: isConn,
    has_cycle: hasCycleResult,
  };
}

export function clearGraph(graphId: string): ClearGraphResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return { success: false };
  }

  graph.nodes.clear();
  graph.adjacency.clear();
  graph.inEdges.clear();

  storeGraph(graph);
  return { success: true };
}

export function cloneGraph(graphId: string): CloneGraphResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return {
      id: '',
      directed: false,
      weighted: false,
      node_count: 0,
      edge_count: 0,
    };
  }

  const newGraphResult = createGraph({
    directed: graph.directed,
    weighted: graph.weighted,
  });

  const newGraph = loadGraph(newGraphResult.id)!;

  // Copy nodes
  for (const node of graph.nodes) {
    newGraph.nodes.add(node);
    newGraph.adjacency.set(node, new Map());
    newGraph.inEdges.set(node, new Map());
  }

  // Copy edges
  for (const [from, neighbors] of graph.adjacency) {
    for (const [to, weight] of neighbors) {
      newGraph.adjacency.get(from)!.set(to, weight);
      newGraph.inEdges.get(to)!.set(from, weight);
    }
  }

  storeGraph(newGraph);
  const edges = getEdges(newGraphResult.id);

  return {
    id: newGraphResult.id,
    directed: graph.directed,
    weighted: graph.weighted,
    node_count: newGraph.nodes.size,
    edge_count: edges.count,
  };
}

export function subgraph(graphId: string, nodeList: string[]): SubgraphResult {
  const graph = loadGraph(graphId);
  if (!graph) {
    return {
      id: '',
      directed: false,
      weighted: false,
      node_count: 0,
      edge_count: 0,
    };
  }

  const newGraphResult = createGraph({
    directed: graph.directed,
    weighted: graph.weighted,
  });

  const newGraph = loadGraph(newGraphResult.id)!;
  const nodeSet = new Set(nodeList);

  // Add nodes that exist in original graph
  for (const node of nodeList) {
    if (graph.nodes.has(node)) {
      newGraph.nodes.add(node);
      newGraph.adjacency.set(node, new Map());
      newGraph.inEdges.set(node, new Map());
    }
  }

  // Add edges between selected nodes
  for (const from of nodeList) {
    const adj = graph.adjacency.get(from);
    if (adj) {
      for (const [to, weight] of adj) {
        if (nodeSet.has(to)) {
          newGraph.adjacency.get(from)!.set(to, weight);
          newGraph.inEdges.get(to)!.set(from, weight);
        }
      }
    }
  }

  storeGraph(newGraph);
  const edges = getEdges(newGraphResult.id);

  return {
    id: newGraphResult.id,
    directed: graph.directed,
    weighted: graph.weighted,
    node_count: newGraph.nodes.size,
    edge_count: edges.count,
  };
}

File: graphlib-2/go/lib/graph.go
package lib

import (
	"sort"
)

// CreateGraph creates a new graph with the given options
func CreateGraph(directed, weighted bool) *Graph {
	g := &Graph{
		ID:       GenerateID(),
		Directed: directed,
		Weighted: weighted,
		Nodes:    make(map[string]bool),
		Adj:      make(map[string]map[string]float64),
		InEdges:  make(map[string]map[string]float64),
	}
	GetStore().Save(g)
	return g
}

// AddNode adds a node to the graph
func (g *Graph) AddNode(nodeID string) (bool, string) {
	g.mu.Lock()
	defer g.mu.Unlock()

	if g.Nodes[nodeID] {
		return false, "node_already_exists"
	}
	g.Nodes[nodeID] = true
	if g.Adj[nodeID] == nil {
		g.Adj[nodeID] = make(map[string]float64)
	}
	if g.Directed && g.InEdges[nodeID] == nil {
		g.InEdges[nodeID] = make(map[string]float64)
	}
	GetStore().Save(g)
	return true, ""
}

// AddEdge adds an edge between two nodes
func (g *Graph) AddEdge(from, to string, weight float64) (bool, string) {
	g.mu.Lock()
	defer g.mu.Unlock()

	// Auto-create nodes if they don't exist
	if !g.Nodes[from] {
		g.Nodes[from] = true
		g.Adj[from] = make(map[string]float64)
		if g.Directed {
			g.InEdges[from] = make(map[string]float64)
		}
	}
	if !g.Nodes[to] {
		g.Nodes[to] = true
		g.Adj[to] = make(map[string]float64)
		if g.Directed {
			g.InEdges[to] = make(map[string]float64)
		}
	}

	// Check if edge already exists
	if _, exists := g.Adj[from][to]; exists {
		return false, "edge_already_exists"
	}

	// For undirected graphs, also check reverse direction
	if !g.Directed {
		if _, exists := g.Adj[to][from]; exists {
			return false, "edge_already_exists"
		}
	}

	// Add the edge
	g.Adj[from][to] = weight
	if g.Directed {
		g.InEdges[to][from] = weight
	} else {
		// For undirected graphs, add edge in both directions
		g.Adj[to][from] = weight
	}

	GetStore().Save(g)
	return true, ""
}

// RemoveNode removes a node and all its incident edges
func (g *Graph) RemoveNode(nodeID string) (bool, string, int) {
	g.mu.Lock()
	defer g.mu.Unlock()

	if !g.Nodes[nodeID] {
		return false, "node_not_found", 0
	}

	removedEdges := 0

	// Count and remove outgoing edges
	if edges, exists := g.Adj[nodeID]; exists {
		removedEdges += len(edges)
		for to := range edges {
			if g.Directed {
				delete(g.InEdges[to], nodeID)
			} else {
				// For undirected, don't double-count self-loops
				if to != nodeID {
					delete(g.Adj[to], nodeID)
				}
			}
		}
	}

	// For directed graphs, count and remove incoming edges
	if g.Directed {
		if edges, exists := g.InEdges[nodeID]; exists {
			for from := range edges {
				if from != nodeID { // Don't double-count self-loops
					removedEdges++
					delete(g.Adj[from], nodeID)
				}
			}
		}
		delete(g.InEdges, nodeID)
	}

	delete(g.Adj, nodeID)
	delete(g.Nodes, nodeID)

	GetStore().Save(g)
	return true, "", removedEdges
}

// RemoveEdge removes an edge between two nodes
func (g *Graph) RemoveEdge(from, to string) (bool, string) {
	g.mu.Lock()
	defer g.mu.Unlock()

	if _, exists := g.Adj[from][to]; !exists {
		// For undirected graphs, check reverse direction
		if !g.Directed {
			if _, exists := g.Adj[to][from]; exists {
				delete(g.Adj[to], from)
				delete(g.Adj[from], to)
				GetStore().Save(g)
				return true, ""
			}
		}
		return false, "edge_not_found"
	}

	delete(g.Adj[from], to)
	if g.Directed {
		delete(g.InEdges[to], from)
	} else {
		delete(g.Adj[to], from)
	}

	GetStore().Save(g)
	return true, ""
}

// GetNodes returns all nodes in the graph
func (g *Graph) GetNodes() []string {
	g.mu.RLock()
	defer g.mu.RUnlock()

	nodes := make([]string, 0, len(g.Nodes))
	for n := range g.Nodes {
		nodes = append(nodes, n)
	}
	sort.Strings(nodes)
	return nodes
}

// GetEdges returns all edges in the graph
func (g *Graph) GetEdges() []Edge {
	g.mu.RLock()
	defer g.mu.RUnlock()

	edges := make([]Edge, 0)
	seen := make(map[string]bool)

	for from := range g.Adj {
		for to, weight := range g.Adj[from] {
			if g.Directed {
				edges = append(edges, Edge{From: from, To: to, Weight: weight})
			} else {
				// For undirected, only add each edge once
				key := from + "-" + to
				reverseKey := to + "-" + from
				if !seen[key] && !seen[reverseKey] {
					seen[key] = true
					edges = append(edges, Edge{From: from, To: to, Weight: weight})
				}
			}
		}
	}

	// Sort edges for deterministic output
	sort.Slice(edges, func(i, j int) bool {
		if edges[i].From != edges[j].From {
			return edges[i].From < edges[j].From
		}
		return edges[i].To < edges[j].To
	})

	return edges
}

// GetNeighbors returns all neighbors of a node
func (g *Graph) GetNeighbors(nodeID string) []string {
	g.mu.RLock()
	defer g.mu.RUnlock()

	neighbors := make([]string, 0)
	if edges, exists := g.Adj[nodeID]; exists {
		for to := range edges {
			neighbors = append(neighbors, to)
		}
	}
	sort.Strings(neighbors)
	return neighbors
}

// HasNode checks if a node exists
func (g *Graph) HasNode(nodeID string) bool {
	g.mu.RLock()
	defer g.mu.RUnlock()
	return g.Nodes[nodeID]
}

// HasEdge checks if an edge exists and returns its weight
func (g *Graph) HasEdge(from, to string) (bool, float64) {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if weight, exists := g.Adj[from][to]; exists {
		return true, weight
	}
	return false, 0
}

// GetDegree returns the degree of a node
func (g *Graph) GetDegree(nodeID string) (int, int, int) {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if !g.Nodes[nodeID] {
		return 0, 0, 0
	}

	outDegree := len(g.Adj[nodeID])

	if g.Directed {
		inDegree := len(g.InEdges[nodeID])
		return inDegree + outDegree, inDegree, outDegree
	}

	// For undirected, in_degree = out_degree = degree
	return outDegree, outDegree, outDegree
}

// NodeCount returns the number of nodes
func (g *Graph) NodeCount() int {
	g.mu.RLock()
	defer g.mu.RUnlock()
	return len(g.Nodes)
}

// EdgeCount returns the number of edges
func (g *Graph) EdgeCount() int {
	g.mu.RLock()
	defer g.mu.RUnlock()

	count := 0
	for _, edges := range g.Adj {
		count += len(edges)
	}

	if !g.Directed {
		count /= 2
	}
	return count
}

// Clear removes all nodes and edges
func (g *Graph) Clear() {
	g.mu.Lock()
	defer g.mu.Unlock()

	g.Nodes = make(map[string]bool)
	g.Adj = make(map[string]map[string]float64)
	g.InEdges = make(map[string]map[string]float64)
	GetStore().Save(g)
}

// Clone creates a copy of the graph
func (g *Graph) Clone() *Graph {
	g.mu.RLock()
	defer g.mu.RUnlock()

	newGraph := &Graph{
		ID:       GenerateID(),
		Directed: g.Directed,
		Weighted: g.Weighted,
		Nodes:    make(map[string]bool),
		Adj:      make(map[string]map[string]float64),
		InEdges:  make(map[string]map[string]float64),
	}

	for n := range g.Nodes {
		newGraph.Nodes[n] = true
	}

	for from, edges := range g.Adj {
		newGraph.Adj[from] = make(map[string]float64)
		for to, weight := range edges {
			newGraph.Adj[from][to] = weight
		}
	}

	if g.Directed {
		for to, edges := range g.InEdges {
			newGraph.InEdges[to] = make(map[string]float64)
			for from, weight := range edges {
				newGraph.InEdges[to][from] = weight
			}
		}
	}

	GetStore().Save(newGraph)
	return newGraph
}

// Subgraph creates a subgraph with only the specified nodes
func (g *Graph) Subgraph(nodeIDs []string) *Graph {
	g.mu.RLock()
	defer g.mu.RUnlock()

	newGraph := &Graph{
		ID:       GenerateID(),
		Directed: g.Directed,
		Weighted: g.Weighted,
		Nodes:    make(map[string]bool),
		Adj:      make(map[string]map[string]float64),
		InEdges:  make(map[string]map[string]float64),
	}

	// Add only the specified nodes that exist
	nodeSet := make(map[string]bool)
	for _, n := range nodeIDs {
		if g.Nodes[n] {
			nodeSet[n] = true
			newGraph.Nodes[n] = true
			newGraph.Adj[n] = make(map[string]float64)
			if g.Directed {
				newGraph.InEdges[n] = make(map[string]float64)
			}
		}
	}

	// Add edges between the selected nodes
	for from := range nodeSet {
		if edges, exists := g.Adj[from]; exists {
			for to, weight := range edges {
				if nodeSet[to] {
					newGraph.Adj[from][to] = weight
					if g.Directed {
						newGraph.InEdges[to][from] = weight
					}
				}
			}
		}
	}

	GetStore().Save(newGraph)
	return newGraph
}

// GetGraphInfo returns metadata about the graph
func (g *Graph) GetGraphInfo() map[string]interface{} {
	g.mu.RLock()
	nodeCount := len(g.Nodes)
	edgeCount := 0
	for _, edges := range g.Adj {
		edgeCount += len(edges)
	}
	if !g.Directed {
		edgeCount /= 2
	}
	g.mu.RUnlock()

	isConnected := g.IsConnected()
	hasCycle, _ := g.HasCycle()

	return map[string]interface{}{
		"id":           g.ID,
		"directed":     g.Directed,
		"weighted":     g.Weighted,
		"node_count":   nodeCount,
		"edge_count":   edgeCount,
		"is_connected": isConnected,
		"has_cycle":    hasCycle,
	}
}
